snakemake/executors.py

```diff
@@ -2052,3 +2059,605 @@ def run_wrapper(
+
+        # Now find (quasi) minimal to satisfy constraints
+        machineTypes = keepers
+        keepers = dict()
+
+        minCores = 1
+        minMem = 3.75
+        smallest = "n1-standard-1"
+
+        for name, machineType in machineTypes.items():
+            if machineType["guestCpus"] < minCores and machineType["memoryMb"] < minMem:
+                smallest = name
+                minCores = machineType["guestCpus"]
+                minMem = machineType["memoryMb"]
+
+        selected = machineTypes[smallest]
+        logger.debug(
+            "Selected machine type {}:{}".format(smallest, selected["description"])
+        )
+        virtualMachine = {
+            "machineType": smallest,
+            "labels": {"app": "snakemake"},
+            "bootDiskSizeGb": 100,  # default is likely 10
+        }
+
+        resources = {"regions": self.regions, "virtualMachine": virtualMachine}
+        return resources
+
+    def _generate_build_package(self):
+        """in order for the instance to access the working directory in storage,
+           we need to upload it. This file is cleaned up at the end of the run.
+           We do this, and then obtain from the instance and extract.
+        """
+        # We will generate a tar.gz package, renamed by hash
+        tmpname = next(tempfile._get_candidate_names())
+        targz = os.path.join(tempfile.gettempdir(), "snakemake-%s.tar.gz" % tmpname)
+        tar = tarfile.open(targz, "w:gz")
+
+        # Add all files to a package
+        for root, dirs, files in os.walk(self.workdir):
+            for filename in files:
+                filename = os.path.join(root, filename)
+                if not filename.endswith("lock") and not filename.startswith(
+                    "snakeworkdir"
+                ):
+                    arcname = filename.replace(self.workdir + os.path.sep, "")
+                    tar.add(filename, arcname=arcname)
+        tar.close()
+
+        # Rename based on hash, in case user wants to save cache
+        sha256 = self.get_file_hash(targz)
+        hash_tar = os.path.join(self.workdir, "snakeworkdir-%s.tar.gz" % sha256)
+
+        # Only copy if we don't have it yet, clean up if we do
+        if not os.path.exists(hash_tar):
+            shutil.move(targz, hash_tar)
+        else:
+            os.remove(targz)
+
+        # We will clean these all up at shutdown
+        self._build_packages.add(hash_tar)
+
+        return hash_tar
+
+    def get_file_hash(self, filename):
+        """find the SHA256 hash string of a file. We use this so that the
+           user can choose to cache working directories in storage.
+        """
+        if os.path.exists(filename):
+            hasher = hashlib.sha256()
+            with open(filename, "rb") as f:
+                for chunk in iter(lambda: f.read(4096), b""):
+                    hasher.update(chunk)
+            return hasher.hexdigest()
+        logger.warning("%s does not exist." % filename)
```

Comment:
This does not need to be a method. It can even go into `snakemake/common.py` for reuse in other places. Also, it should fail (with FileNotFoundError as coming from open) in case of non-existence I think.

Reply:
okay can do.